hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/listener/ClientSmartListenerService.java

```diff
@@ -74,7 +77,7 @@ public String call() {
                         invoke(registrationKey, connection);
                     }
                 } catch (Exception e) {
-                    deregisterListener(userRegistrationId);
```

Comment:
Understand the deadlock reason. The change introduced a behaviour difference. The registerListener call will now block until a cleanup is done (deregister finishes) in case of failure, and sometimes this cleanup may take some time since the server may not respond. Is this the behaviour we want?